File: tests/test_auth/hosted_group_payload.py
"""Execute the shipped client with a fake browser/transport; no network or tokens."""
import json
from pathlib import Path
import shutil
import subprocess

from testit import helpers as th

TESTIT_TIER = "bug"
ROOT = Path(__file__).resolve().parents[2]


def _payloads(method, branded):
    node = shutil.which("node")
    assert node, "Node.js is required for the hosted-auth client regression"
    result = subprocess.run(
        [node, str(Path(__file__).with_suffix(".js")),
         str(ROOT / "mojo/apps/account/static/account/mojo-auth.js"), method,
         "brand" if branded else "plain"],
        capture_output=True, text=True, timeout=20,
    )
    assert result.returncode == 0, result.stderr
    return json.loads(result.stdout)


def _check(method, branded):
    calls = _payloads(method, branded)
    assert len(calls) == (1 if method == "sms" else 2), calls
    for call in calls:
        if branded:
            assert call["body"].get("group_uuid") == "test-brand-4622", call
        else:
            assert "group_uuid" not in call["body"], call
    complete = calls[-1]["body"]
    assert complete["duid"] == "test-device", complete
    assert complete["bouncer_token"] == "test-bouncer", complete
    if method == "sms":
        assert complete["username"] == "test-user" and complete["code"] == "123456"
    else:
        assert complete["challenge_id"] == "test-challenge"
        assert complete["credential"]["response"]["signature"] == "AQ"
        assert ("username" in calls[0]["body"]) == (method == "named")


@th.django_unit_test("hosted discoverable passkey carries group on begin and complete")
def test_hosted_discoverable_group(opts):
    _check("discoverable", True)


@th.django_unit_test("hosted username passkey carries group on begin and complete")
def test_hosted_named_group(opts):
    _check("named", True)


@th.django_unit_test("hosted SMS verification carries group")
def test_hosted_sms_group(opts):
    _check("sms", True)


@th.django_unit_test("hosted auth helpers preserve calls without options")
def test_hosted_legacy_options(opts):
    for method in ("discoverable", "named", "sms"):
        _check(method, False)


@th.django_unit_test("hosted login template supplies resolved group to both completions")
def test_hosted_template_group(opts):
    template = (ROOT / "mojo/apps/account/templates/account/login.html").read_text()
    assert "loginWithPasskeyDiscoverable({ group_uuid: cfg.groupUuid })" in template
    assert "verifySmsLogin(phone, code, { group_uuid: cfg.groupUuid })" in template

File: tests/test_auth/hosted_group_payload.js
// Run the real client in a browser-shaped VM, recording its outgoing JSON.
const fs = require('node:fs');
const vm = require('node:vm');
const assert = require('node:assert/strict');
const calls = [];
const stored = new Map();
const bytes = new Uint8Array([1]).buffer;
const context = {
    console, atob, btoa, Uint8Array,
    localStorage: {
        getItem: key => stored.get(key) || null,
        setItem: (key, value) => stored.set(key, value),
        removeItem: key => stored.delete(key)
    },
    window: {
        PublicKeyCredential: function () {},
        MojoBouncer: { getToken: () => 'test-bouncer', getDuid: () => 'test-device' }
    },
    navigator: { credentials: { get: async () => ({
        id: 'credential', rawId: bytes, type: 'public-key',
        response: { clientDataJSON: bytes, authenticatorData: bytes,
                    signature: bytes, userHandle: null }
    }) } },
    fetch: async (url, options) => {
        calls.push({ url, body: JSON.parse(options.body) });
        const data = url.endsWith('/begin')
            ? { challenge_id: 'test-challenge', publicKey: { challenge: 'AQ', allowCredentials: [] } }
            : { access_token: 'synthetic-access', refresh_token: 'synthetic-refresh' };
        return { ok: true, json: async () => ({ status: true, data }) };
    }
};
vm.createContext(context);
vm.runInContext(fs.readFileSync(process.argv[2], 'utf8'), context);
(async () => {
    const auth = context.MojoAuth;
    auth.init({ baseURL: 'https://auth.example.test' });
    const options = process.argv[4] === 'brand' ? { group_uuid: 'test-brand-4622' } : undefined;
    if (process.argv[3] === 'sms') await auth.verifySmsLogin('test-user', '123456', options);
    else if (process.argv[3] === 'named') await auth.loginWithPasskey('test-user', options);
    else await auth.loginWithPasskeyDiscoverable(options);
    assert.equal(stored.get('access_token'), 'synthetic-access');
    assert.equal(stored.get('refresh_token'), 'synthetic-refresh');
    process.stdout.write(JSON.stringify(calls));
})().catch(error => { console.error(error); process.exitCode = 1; });
